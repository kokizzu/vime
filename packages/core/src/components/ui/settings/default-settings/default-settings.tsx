import {
  h, Component, Prop, Watch, forceUpdate,
} from '@stencil/core';
import { withPlayerContext } from '../../../core/player/PlayerContext';
import { PlayerProps } from '../../../core/player/PlayerProps';
import { Disposal } from '../../../core/player/Disposal';
import { listen } from '../../../../utils/dom';
import { isUndefined } from '../../../../utils/unit';
import { Dispatcher, createDispatcher } from '../../../core/player/PlayerDispatcher';
import { findRootPlayer } from '../../../core/player/utils';

/**
 * @slot - Used to extend the settings with additional menu options (see `vime-submenu` or
 * `vime-menu-item`).
 */
@Component({
  tag: 'vime-default-settings',
})
export class DefaultSettings {
  private textTracksDisposal = new Disposal();

  private dispatch!: Dispatcher;

  private player?: HTMLVimePlayerElement;

  private rateSubmenu: any;

  private qualitySubmenu: any;

  private captionsSubmenu: any;

  /**
   * @internal
   */
  @Prop() i18n: PlayerProps['i18n'] = {};

  /**
   * @internal
   */
  @Prop() playbackRate: PlayerProps['playbackRate'] = 1;

  /**
   * @internal
   */
  @Prop() playbackRates: PlayerProps['playbackRates'] = [1];

  /**
   * @internal
   */
  @Prop() playbackQuality?: PlayerProps['playbackQuality'];

  /**
   * @internal
   */
  @Prop() playbackQualities: PlayerProps['playbackQualities'] = [];

  /**
   * @internal
   */
  @Prop() isCaptionsActive: PlayerProps['isCaptionsActive'] = false;

  /**
   * @internal
   */
  @Prop() currentCaption?: PlayerProps['currentCaption'];

  /**
   * @internal
   */
  @Prop() textTracks?: PlayerProps['textTracks'];

  @Watch('textTracks')
  onTextTracksChange() {
    this.textTracksDisposal.empty();
    if (isUndefined(this.textTracks)) return;
    this.textTracksDisposal.add(listen(this.textTracks!, 'change', () => {
      setTimeout(() => forceUpdate(this), 300);
    }));
  }

  connectedCallback() {
    this.skipFirstRender = true;
    this.player = findRootPlayer(this);
    this.dispatch = createDispatcher(this);
  }

  private skipFirstRender = true;

  componentWillRender() {
    if (this.skipFirstRender) {
      this.skipFirstRender = false;
      return undefined;
    }

    return Promise.all([
      this.buildPlaybackRateSubmenu(),
      this.buildPlaybackQualitySubmenu(),
      this.buildCaptionsSubmenu(),
    ]);
  }

  disconnectedCallback() {
    this.player = undefined;
    this.textTracksDisposal.empty();
  }

  private onPlaybackRateSelect(event: Event) {
    const radio = event.target as HTMLVimeMenuRadioElement;
    this.dispatch('playbackRate', parseFloat(radio.value));
  }

  private async buildPlaybackRateSubmenu() {
    const canSetPlaybackRate = await this.player?.canSetPlaybackRate();

    if (this.playbackRates.length === 1 || !canSetPlaybackRate) {
      this.rateSubmenu = (
        <vime-menu-item label={this.i18n.playbackRate} hint={this.i18n.normal} />
      );
      return;
    }

    const formatRate = (rate: number) => ((rate === 1) ? this.i18n.normal : `${rate}`);

    const radios = this.playbackRates.map((rate) => (
      <vime-menu-radio
        label={formatRate(rate)}
        value={`${rate}`}
      />
    ));

    this.rateSubmenu = (
      <vime-submenu label={this.i18n.playbackRate} hint={formatRate(this.playbackRate)}>
        <vime-menu-radio-group
          value={`${this.playbackRate}`}
          onVCheck={this.onPlaybackRateSelect.bind(this)}
        >
          {radios}
        </vime-menu-radio-group>
      </vime-submenu>
    );
  }

  private onPlaybackQualitySelect(event: Event) {
    const radio = event.target as HTMLVimeMenuRadioElement;
    this.dispatch('playbackQuality', radio.value);
  }

  private async buildPlaybackQualitySubmenu() {
    const canSetPlaybackQuality = await this.player?.canSetPlaybackQuality();

    if (this.playbackQualities.length === 0 || !canSetPlaybackQuality) {
      this.qualitySubmenu = (
        <vime-menu-item
          label={this.i18n.playbackQuality}
          hint={this.playbackQuality ?? this.i18n.auto}
        />
      );
      return;
    }

    // @TODO this doesn't account for audio qualities yet.
    const getBadge = (quality: string) => {
      const verticalPixels = parseInt(quality.slice(0, -1), 10);
      if (verticalPixels > 2160) return 'UHD';
      if (verticalPixels >= 1080) return 'HD';
      return undefined;
    };

    const radios = this.playbackQualities.map((quality) => (
      <vime-menu-radio
        label={quality}
        value={quality}
        badge={getBadge(quality)}
      />
    ));

    this.qualitySubmenu = (
      <vime-submenu label={this.i18n.playbackQuality} hint={this.playbackQuality}>
        <vime-menu-radio-group
          value={this.playbackQuality}
          onVCheck={this.onPlaybackQualitySelect.bind(this)}
        >
          {radios}
        </vime-menu-radio-group>
      </vime-submenu>
    );
  }

  private async onCaptionSelect(event: Event) {
    const radio = event.target as HTMLVimeMenuRadioElement;
    const index = parseInt(radio.value, 10);
    const player = findRootPlayer(this);

    if (index === -1) {
      await player.toggleCaptionsVisibility(false);
      return;
    }

    const track = Array.from(this.textTracks ?? [])[index];
    if (!isUndefined(track)) {
      if (!isUndefined(this.currentCaption)) this.currentCaption!.mode = 'disabled';
      track.mode = 'showing';
      await player.toggleCaptionsVisibility(true);
    }
  }

  private async buildCaptionsSubmenu() {
    const captions = Array.from(this.textTracks ?? [])
      .filter((track) => ['captions', 'subtitles'].includes(track.kind));

    if (captions.length === 0) {
      this.captionsSubmenu = (
        <vime-menu-item label={this.i18n.subtitlesOrCc} hint={this.i18n.none} />
      );
      return;
    }

    const getTrackValue = (
      track: TextTrack,
    ) => `${Array.from(this.textTracks!).findIndex((t) => t === track)}`;

    const radios = [(
      <vime-menu-radio
        label={this.i18n.off}
        value="-1"
      />
    )].concat(captions.map((track) => (
      <vime-menu-radio
        label={track.label}
        value={getTrackValue(track)}
      />
    )));

    const groupValue = (!this.isCaptionsActive || isUndefined(this.currentCaption))
      ? '-1'
      : getTrackValue(this.currentCaption!);

    this.captionsSubmenu = (
      <vime-submenu
        label={this.i18n.subtitlesOrCc}
        hint={(this.isCaptionsActive ? this.currentCaption?.label : undefined) ?? this.i18n.off}
      >
        <vime-menu-radio-group
          value={groupValue}
          onVCheck={this.onCaptionSelect.bind(this)}
        >
          {radios}
        </vime-menu-radio-group>
      </vime-submenu>
    );
  }

  render() {
    return (
      <vime-settings>
        {this.rateSubmenu}
        {this.qualitySubmenu}
        {this.captionsSubmenu}
        <slot />
      </vime-settings>
    );
  }
}

withPlayerContext(DefaultSettings, [
  'i18n',
  'playbackRate',
  'playbackRates',
  'playbackQuality',
  'playbackQualities',
  'isCaptionsActive',
  'currentCaption',
  'textTracks',
]);
