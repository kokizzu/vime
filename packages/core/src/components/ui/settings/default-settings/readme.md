# vime-default-settings

Creates a settings menu with options for changing the playback rate, quality and captions of
the current media. This component is provider aware. For example, it will only show options for
changing the playback rate if the current provider allows changing it (`player.canSetPlaybackRate()`).
In addition, you can extend the settings with more options via the default `slot`.

## Visual

<img
  src="https://raw.githubusercontent.com/vime-js/vime/master/packages/core/src/components/ui/settings/default-settings/default-settings.png"
  alt="Vime default settings component"
/>

<!-- Auto Generated Below -->

## Usage

### Angular

```html {5} title="example.html"
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
    <vime-default-settings></vime-default-settings>
  </vime-ui>
</vime-player>
```

### Html

```html {5}
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
    <vime-default-settings></vime-default-settings>
  </vime-ui>
</vime-player>
```

### React

```tsx {2,10}
import React from 'react';
import { VimePlayer, VimeUi, VimeDefaultSettings } from '@vime/react';

function Example() {
  return render(
    <VimePlayer>
      {/* ... */}
      <VimeUi>
        {/* ... */}
        <VimeDefaultSettings />
      </VimeUi>
    </VimePlayer>
  );
}
```

### Vue

```html {6,12,18} title="example.vue"
<template>
  <VimePlayer>
    <!-- ... -->
    <VimeUi>
      <!-- ... -->
      <VimeDefaultSettings />
    </VimeUi>
  </VimePlayer>
</template>

<script>
  import { VimePlayer, VimeUi, VimeDefaultSettings } from '@vime/vue';

  export default {
    components: {
      VimePlayer,
      VimeUi,
      VimeDefaultSettings,
    },
  };
</script>
```

## Slots

| Slot | Description                                                                                        |
| ---- | -------------------------------------------------------------------------------------------------- |
|      | Used to extend the settings with additional menu options (see `vime-submenu` or `vime-menu-item`). |

## Dependencies

### Used by

- [vime-default-ui](../../default-ui)

### Depends on

- [vime-menu-item](../menu-item)
- [vime-menu-radio](../menu-radio)
- [vime-submenu](../submenu)
- [vime-menu-radio-group](../menu-radio-group)
- [vime-settings](../settings)

### Graph

```mermaid
graph TD;
  vime-default-settings --> vime-menu-item
  vime-default-settings --> vime-menu-radio
  vime-default-settings --> vime-submenu
  vime-default-settings --> vime-menu-radio-group
  vime-default-settings --> vime-settings
  vime-menu-item --> vime-icon
  vime-menu-radio --> vime-menu-item
  vime-submenu --> vime-menu-item
  vime-submenu --> vime-menu
  vime-settings --> vime-menu
  vime-default-ui --> vime-default-settings
  style vime-default-settings fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
