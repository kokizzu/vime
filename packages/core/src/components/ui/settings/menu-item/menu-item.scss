vime-menu-item {
  /**
   * @prop --menu-item-hint-font-size: The font size of the hint text.
   * @prop --menu-item-hint-color: The color of the hint text.
   * @prop --menu-item-hint-opacity: The opacity of the hint text.
   * @prop --menu-item-badge-color: The color of the badge text.
   * @prop --menu-item-badge-bg: The background color of the badge.
   * @prop --menu-item-badge-font-size: The font size of the the badge text. 
   * @prop --menu-item-arrow-color: The color of the left/right arrow on a navigational menu item.
   * @prop --menu-item-check-icon-width: The width of the checked icon.
   * @prop --menu-item-check-icon-height: The height of the checked icon.
   * @prop --menu-item-padding: The padding within each menu item.
   * @prop --menu-item-focus-color: The color of a menu item's text when it is being focused or hovered on.
   * @prop --menu-item-focus-bg: The background color of a menu item when it is being focused or hovered on.
   * @prop --menu-item-tap-highlight: The highlight color when a menu item is tapped.
   * @prop --menu-item-divider-color: The color of the divider between a navigational menu item and the menu body.
   */

  display: flex;
  align-items: center;
  flex-direction: row;
  width: 100%;
  cursor: pointer;
  color: var(--menu-color);
  background: var(--menu-bg);
  font-size: var(--menu-font-size);
  font-weight: var(--menu-font-weight);
  padding: var(--menu-item-padding);

  &:focus {
    outline: 0;
  }

  &.hidden {
    display: none;
  }

  &.tapHighlight {
    background: var(--menu-item-tap-highlight);
  }

  &.showDivider {
    border-bottom: 0.5px solid var(--menu-item-divider-color);
  }

  &.notTouch {
    &:hover,
    &:focus {
      outline: 0;
      color: var(--menu-item-focus-color);
      background-color: var(--menu-item-focus-bg);
    }
  }

  &[aria-hidden='true'] {
    display: none;
  }

  vime-icon {
    display: inline-block;
  }

  svg {
    fill: currentColor;
    pointer-events: none;
    width: var(--menu-item-check-icon-width, 10px) !important;
    height: var(--menu-item-check-icon-height, 10px) !important;
    margin-right: 10px;
    @include invisible();
  }

  &[aria-checked='true'] svg {
    @include visible();
  }

  .hint {
    display: inline-block;
    margin-left: auto;
    overflow: hidden;
    pointer-events: none;
    margin-right: 6px;
    font-size: var(--menu-item-hint-font-size);
    opacity: var(--menu-item-hint-opacity);
    color: var(--menu-item-hint-color);
  }
  
  .badge {
    display: inline-block;
    line-height: 1;
    overflow: hidden;
    pointer-events: none;
    color: var(--menu-item-badge-color);
    background: var(--menu-item-badge-bg);
    font-size: var(--menu-item-badge-font-size);
  }

  .spacer {
    flex: 1;
  }

  .arrow {
    color: var(--menu-item-arrow-color);
    border: 2px solid;
    padding: 2px;
    display: inline-block;
    border-width: 0 2px 2px 0;
  
    &.left {
      margin-right: 6px;
      transform: rotate(135deg);
    }
  
    &.right {
      transform: rotate(-45deg);
      opacity: 0.38;
    }
  }
}
