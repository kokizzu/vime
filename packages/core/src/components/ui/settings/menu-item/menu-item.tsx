/* eslint-disable jsx-a11y/no-noninteractive-tabindex */
/* eslint-disable jsx-a11y/role-supports-aria-props */

import {
  h, Host, Component, Prop, State, Element,
} from '@stencil/core';
import { isUndefined, isNull } from '../../../../utils/unit';
import { PlayerProps, PlayerProp } from '../../../core/player/PlayerProp';
import { openPlayerWormhole } from '../../../core/player/PlayerWormhole';

@Component({
  tag: 'vime-menu-item',
  styleUrl: 'menu-item.scss',
})
export class MenuItem {
  @Element() el!: HTMLVimeMenuItemElement;

  @State() showTapHighlight = false;

  /**
   * The `id` attribute of the item.
   */
  @Prop() identifier?: string;

  /**
   * Whether the item is displayed or not.
   */
  @Prop() hidden = false;

  /**
   * The label/title of the item.
   */
  @Prop() label!: string;

  /**
   * If the item has a popup menu, then this should be the `id` of said menu. Sets the
   * `aria-controls` property.
   */
  @Prop() menu?: string;

  /**
   * If the item has a popup menu, this indicates whether the menu is open or not. Sets the
   * `aria-expanded` property.
   */
  @Prop() expanded?: boolean;

  /**
   * If this item is to behave as a radio button, then this property determines whether the
   * radio is selected or not. Sets the `aria-checked` property.
   */
  @Prop() checked?: boolean;

  /**
   * This can provide additional context about some underlying state of the item. For example, if
   * the menu item opens/closes a submenu with options, the hint could be the currently selected
   * option.
   */
  @Prop() hint?: string;

  /**
   * This can provide additional context about the value of a menu item. For example, if the item
   * is a radio button for a set of video qualities, the badge could describe whether the quality
   * is UHD, HD etc.
   */
  @Prop() badge?: string;

  /**
   * The URL to an SVG element or fragment to load.
   */
  @Prop() checkedIcon?: string = '#vime-checkmark';

  /**
   * @internal
   */
  @Prop() isTouch: PlayerProps[PlayerProp.IsTouch] = false;

  private onClick() {
    if (isUndefined(this.menu)) return;
    const submenu = document.querySelector(`#${this.menu}`) as HTMLVimeMenuElement;
    if (!isNull(submenu)) submenu!.active = !this.expanded;
  }

  private onTouchStart() {
    this.showTapHighlight = true;
    setTimeout(() => { this.showTapHighlight = false; }, 100);
  }

  private onMouseLeave() {
    this.el.blur();
  }

  render() {
    const isCheckedDefined = !isUndefined(this.checked);
    const isMenuDefined = !isUndefined(this.menu);
    const hasExpanded = this.expanded ? 'true' : 'false';
    const isChecked = this.checked ? 'true' : 'false';
    const showCheckedIcon = isCheckedDefined && !isUndefined(this.checkedIcon);
    const showLeftNavArrow = isMenuDefined && this.expanded;
    const showRightNavArrow = isMenuDefined && !this.expanded;
    const showHint = !isUndefined(this.hint)
      && !isCheckedDefined
      && (!isMenuDefined || !this.expanded);
    const showBadge = !isUndefined(this.badge) && (!showHint && !showRightNavArrow);
    const hasSpacer = showHint || showBadge || showRightNavArrow;

    return (
      <Host
        class={{
          notTouch: !this.isTouch,
          tapHighlight: this.showTapHighlight,
          showDivider: isMenuDefined && (this.expanded ?? false),
        }}
        id={this.identifier}
        role={isCheckedDefined ? 'menuitemradio' : 'menuitem'}
        tabindex="0"
        aria-label={this.label}
        aria-hidden={this.hidden ? 'true' : 'false'}
        aria-haspopup={isMenuDefined ? 'true' : undefined}
        aria-controls={this.menu}
        aria-expanded={isMenuDefined ? hasExpanded : undefined}
        aria-checked={isCheckedDefined ? isChecked : undefined}
        onClick={this.onClick.bind(this)}
        onTouchStart={this.onTouchStart.bind(this)}
        onMouseLeave={this.onMouseLeave.bind(this)}
      >
        {showCheckedIcon && <vime-icon href={this.checkedIcon!} />}
        {showLeftNavArrow && <span class="arrow left" />}
        {this.label}
        {hasSpacer && <span class="spacer" />}
        {showHint && <span class="hint">{this.hint}</span>}
        {showBadge && <span class="badge">{this.badge}</span>}
        {showRightNavArrow && <span class="arrow right" />}
      </Host>
    );
  }
}

openPlayerWormhole(MenuItem, [
  PlayerProp.IsTouch,
]);
