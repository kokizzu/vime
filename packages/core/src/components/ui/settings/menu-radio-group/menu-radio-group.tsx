import {
  h, Component, Element, Prop, Watch, Event, EventEmitter, Listen,
} from '@stencil/core';

/**
 * @slot - Used to pass in radio buttons (`vime-menu-radio`).
 */
@Component({
  tag: 'vime-menu-radio-group',
})
export class MenuRadioGroup {
  @Element() el!: HTMLVimeMenuRadioGroupElement;

  /**
   * The current value selected for this group.
   */
  @Prop({ mutable: true }) value?: string;

  @Watch('value')
  onValueChange() {
    this.findRadios().forEach((radio) => {
      // eslint-disable-next-line no-param-reassign
      radio.checked = (radio.value === this.value);
    });
  }

  /**
   * Emitted when a new radio button is selected for this group.
   */
  @Event() vCheck!: EventEmitter<void>;

  connectedCallback() {
    this.onValueChange();
  }

  @Listen('vCheck')
  onSelectionChange(event: Event) {
    const radio = event.target as HTMLVimeMenuRadioElement;
    this.value = radio.value;
  }

  private findRadios() {
    return this.el.querySelectorAll('vime-menu-radio');
  }

  render() {
    return (
      <slot />
    );
  }
}
