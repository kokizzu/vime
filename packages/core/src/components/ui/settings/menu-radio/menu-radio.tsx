/* eslint-disable jsx-a11y/label-has-associated-control */

import {
  h, Component, Prop, Event, EventEmitter,
} from '@stencil/core';

@Component({
  tag: 'vime-menu-radio',
})
export class MenuRadio {
  /**
   * The title of the radio item displayed to the user.
   */
  @Prop() label!: string;

  /**
   * The value associated with this radio item.
   */
  @Prop() value!: string;

  /**
   * Whether the radio item is selected or not.
   */
  @Prop({ mutable: true }) checked = false;

  /**
   * This can provide additional context about the value. For example, if the option is for a set of
   * video qualities, the badge could describe whether the quality is UHD, HD etc.
   */
  @Prop() badge?: string;

  /**
   * The URL to an SVG element or fragment to load.
   */
  @Prop() checkedIcon?: string = '#vime-checkmark';

  /**
   * Emitted when the radio button is selected.
   */
  @Event() check!: EventEmitter<void>;

  private onClick() {
    this.checked = !this.checked;
    this.check.emit();
  }

  render() {
    return (
      <vime-menu-item
        label={this.label}
        checked={this.checked}
        badge={this.badge}
        checkedIcon={this.checkedIcon}
        onClick={this.onClick.bind(this)}
      />
    );
  }
}
