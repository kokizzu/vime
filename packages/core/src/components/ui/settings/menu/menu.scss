vime-menu {
  /**
   * @prop --menu-color: The text color within the menu.
   * @prop --menu-bg: The background color the menu.
   * @prop --menu-font-size: The font size of text within the menu.
   * @prop --menu-font-weight: The font weight of text within the menu.
   */

  display: flex;
  flex-direction: column;
  position: relative;
  overflow: hidden;
  width: 100%;
  height: auto;
  pointer-events: auto;
  color: var(--menu-color);
  background: var(--menu-bg);
  font-size: var(--menu-font-size);
  font-weight: var(--menu-font-weight);
  text-align: left;
  z-index: $z-index-menu;

  &[aria-hidden='true'] {
    display: none;
  }

  &:focus {
    outline: 0;
  }
}