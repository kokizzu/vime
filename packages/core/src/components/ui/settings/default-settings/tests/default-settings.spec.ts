import { SpecPage } from '@stencil/core/testing';
import { DefaultSettings } from '../default-settings';
import { newUISpecPage } from '../../../ui/tests';
import { Settings } from '../../settings/settings';
import { MenuRadioGroup } from '../../menu-radio-group/menu-radio-group';
import { MenuRadio } from '../../menu-radio/menu-radio';
import { MenuItem } from '../../menu-item/menu-item';
import { Menu } from '../../menu/menu';
import { Submenu } from '../../submenu/submenu';
import { MockMediaProviderAdapter } from '../../../../providers/MediaProvider';

let page: SpecPage;
let player: HTMLVimePlayerElement;
let provider: HTMLVimeFaketubeElement;
let defaultSettings: HTMLVimeDefaultSettingsElement;

const getSettingsMenu = () => page.root!.querySelector('vime-settings');

const getController = (label: string) => page.root!
  .querySelector(`vime-menu-item[aria-label="${label}"]`)! as HTMLVimeMenuItemElement;

const getPlaybackRateSubmenuController = () => getController('Playback Rate');
const getPlaybackQualitySubmenuController = () => getController('Playback Quality');

beforeEach(async () => {
  ({ page } = await newUISpecPage(
    [
      DefaultSettings, Settings, Menu,
      Submenu, MenuRadioGroup, MenuRadio,
      MenuItem,
    ],
    '<vime-default-settings>DefaultSlot</vime-default-settings>',
  ));

  player = page.root! as HTMLVimePlayerElement;
  provider = page.root!.querySelector('vime-faketube')!;
  defaultSettings = page.root!.querySelector('vime-default-settings')!;

  // Trigger another render cycle for the submenus to be built.
  defaultSettings.playbackRates = [1];
  await page.waitForChanges();
});

it('should be structurally sound', () => {
  expect(page.root).toMatchSnapshot();
});

test('playback rate submenu should be structurally sound', async () => {
  const adapter = await player.getAdapter() as MockMediaProviderAdapter;
  adapter.canSetPlaybackRate!.mockReturnValue(Promise.resolve(true));
  provider.dispatchChange('playbackRates', [1, 2, 3]);
  await page.waitForChanges();
  getPlaybackRateSubmenuController().click();
  await page.waitForChanges();
  expect(getSettingsMenu()).toMatchSnapshot();
});

it('should change playback rate on radio selection', async () => {
  const adapter = await player.getAdapter() as MockMediaProviderAdapter;
  adapter.canSetPlaybackRate!.mockReturnValue(Promise.resolve(true));
  provider.dispatchChange('playbackRates', [1, 2, 3]);
  await page.waitForChanges();
  getPlaybackRateSubmenuController().click();
  await page.waitForChanges();
  const radio = page.root!.querySelector('vime-menu-item[aria-label="2"]')! as HTMLVimeMenuItemElement;
  radio.click();
  await page.waitForChanges();
  expect(player.playbackRate).toEqual(2);
});

it('should update selected radio on playback rate change', async () => {
  const adapter = await player.getAdapter() as MockMediaProviderAdapter;
  adapter.canSetPlaybackRate!.mockReturnValue(Promise.resolve(true));
  provider.dispatchChange('playbackRates', [1, 2, 3]);
  await page.waitForChanges();
  getPlaybackRateSubmenuController().click();
  await page.waitForChanges();
  player.playbackRate = 3;
  await page.waitForChanges();
  const radio = page.root!.querySelector('vime-menu-item[aria-label="3"]')! as HTMLVimeMenuItemElement;
  expect(radio.checked).toBeTruthy();
});

test('playback quality submenu should be structurally sound', async () => {
  const adapter = await player.getAdapter() as MockMediaProviderAdapter;
  adapter.canSetPlaybackQuality!.mockReturnValue(Promise.resolve(true));
  provider.dispatchChange('playbackQuality', '720p');
  provider.dispatchChange('playbackQualities', ['1080p', '720p', '480p']);
  await page.waitForChanges();
  getPlaybackQualitySubmenuController().click();
  await page.waitForChanges();
  expect(getSettingsMenu()).toMatchSnapshot();
});

it('should change playback quality on radio selection', async () => {
  const adapter = await player.getAdapter() as MockMediaProviderAdapter;
  adapter.canSetPlaybackQuality!.mockReturnValue(Promise.resolve(true));
  provider.dispatchChange('playbackQuality', '720p');
  provider.dispatchChange('playbackQualities', ['1080p', '720p', '480p']);
  await page.waitForChanges();
  getPlaybackQualitySubmenuController().click();
  await page.waitForChanges();
  const radio = page.root!.querySelector('vime-menu-item[aria-label="1080p"]')! as HTMLVimeMenuItemElement;
  radio.click();
  await page.waitForChanges();
  expect(player.playbackQuality).toEqual('1080p');
});

it('should update selected radio on playback quality change', async () => {
  const adapter = await player.getAdapter() as MockMediaProviderAdapter;
  adapter.canSetPlaybackQuality!.mockReturnValue(Promise.resolve(true));
  provider.dispatchChange('playbackQuality', '720p');
  provider.dispatchChange('playbackQualities', ['1080p', '720p', '480p']);
  await page.waitForChanges();
  getPlaybackQualitySubmenuController().click();
  await page.waitForChanges();
  player.playbackQuality = '480p';
  await page.waitForChanges();
  const radio = page.root!.querySelector('vime-menu-item[aria-label="480p"]')! as HTMLVimeMenuItemElement;
  expect(radio.checked).toBeTruthy();
});
