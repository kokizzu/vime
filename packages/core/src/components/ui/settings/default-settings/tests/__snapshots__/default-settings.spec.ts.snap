// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`playback quality submenu should be structurally sound 1`] = `
<vime-settings style="bottom: 73px;">
  <!---->
  <vime-menu aria-hidden="true" aria-labelledby="" id="vime-settings-5" role="menu" tabindex="-1">
    <!---->
    <div>
      <vime-menu-item aria-hidden="false" aria-label="Playback Rate" class="notTouch" role="menuitem" tabindex="0">
        Playback Rate
        <span class="spacer"></span>
        <span class="hint">
          Normal
        </span>
      </vime-menu-item>
      <vime-submenu>
        <!---->
        <vime-menu-item aria-controls="vime-submenu-4" aria-expanded="false" aria-haspopup="true" aria-hidden="false" aria-label="Playback Quality" class="notTouch" id="vime-submenu-4-controller" role="menuitem" tabindex="0">
          Playback Quality
          <span class="spacer"></span>
          <span class="hint">
            720p
          </span>
          <span class="arrow right"></span>
        </vime-menu-item>
        <vime-menu aria-hidden="true" aria-labelledby="vime-submenu-4-controller" id="vime-submenu-4" role="menu" tabindex="-1">
          <!---->
          <div>
            <vime-menu-radio-group>
              <!---->
              <vime-menu-radio>
                <vime-menu-item aria-checked="false" aria-hidden="false" aria-label="1080p" class="notTouch" role="menuitemradio" tabindex="0">
                  <vime-icon href="#vime-checkmark"></vime-icon>
                  1080p
                  <span class="spacer"></span>
                  <span class="badge">
                    HD
                  </span>
                </vime-menu-item>
              </vime-menu-radio>
              <vime-menu-radio>
                <vime-menu-item aria-checked="true" aria-hidden="false" aria-label="720p" class="notTouch" role="menuitemradio" tabindex="0">
                  <vime-icon href="#vime-checkmark"></vime-icon>
                  720p
                </vime-menu-item>
              </vime-menu-radio>
              <vime-menu-radio>
                <vime-menu-item aria-checked="false" aria-hidden="false" aria-label="480p" class="notTouch" role="menuitemradio" tabindex="0">
                  <vime-icon href="#vime-checkmark"></vime-icon>
                  480p
                </vime-menu-item>
              </vime-menu-radio>
            </vime-menu-radio-group>
          </div>
        </vime-menu>
      </vime-submenu>
      <vime-menu-item aria-hidden="false" aria-label="Subtitles/CC" class="notTouch" role="menuitem" tabindex="0">
        Subtitles/CC
        <span class="spacer"></span>
        <span class="hint">
          None
        </span>
      </vime-menu-item>
      DefaultSlot
    </div>
  </vime-menu>
</vime-settings>
`;

exports[`playback rate submenu should be structurally sound 1`] = `
<vime-settings style="bottom: 73px;">
  <!---->
  <vime-menu aria-hidden="true" aria-labelledby="" id="vime-settings-2" role="menu" tabindex="-1">
    <!---->
    <div>
      <vime-submenu>
        <!---->
        <vime-menu-item aria-controls="vime-submenu-1" aria-expanded="false" aria-haspopup="true" aria-hidden="false" aria-label="Playback Rate" class="notTouch" id="vime-submenu-1-controller" role="menuitem" tabindex="0">
          Playback Rate
          <span class="spacer"></span>
          <span class="hint">
            Normal
          </span>
          <span class="arrow right"></span>
        </vime-menu-item>
        <vime-menu aria-hidden="true" aria-labelledby="vime-submenu-1-controller" id="vime-submenu-1" role="menu" tabindex="-1">
          <!---->
          <div>
            <vime-menu-radio-group>
              <!---->
              <vime-menu-radio>
                <vime-menu-item aria-checked="true" aria-hidden="false" aria-label="Normal" class="notTouch" role="menuitemradio" tabindex="0">
                  <vime-icon href="#vime-checkmark"></vime-icon>
                  Normal
                </vime-menu-item>
              </vime-menu-radio>
              <vime-menu-radio>
                <vime-menu-item aria-checked="false" aria-hidden="false" aria-label="2" class="notTouch" role="menuitemradio" tabindex="0">
                  <vime-icon href="#vime-checkmark"></vime-icon>
                  2
                </vime-menu-item>
              </vime-menu-radio>
              <vime-menu-radio>
                <vime-menu-item aria-checked="false" aria-hidden="false" aria-label="3" class="notTouch" role="menuitemradio" tabindex="0">
                  <vime-icon href="#vime-checkmark"></vime-icon>
                  3
                </vime-menu-item>
              </vime-menu-radio>
            </vime-menu-radio-group>
          </div>
        </vime-menu>
      </vime-submenu>
      <vime-menu-item aria-hidden="false" aria-label="Playback Quality" class="notTouch" role="menuitem" tabindex="0">
        Playback Quality
        <span class="spacer"></span>
        <span class="hint">
          Auto
        </span>
      </vime-menu-item>
      <vime-menu-item aria-hidden="false" aria-label="Subtitles/CC" class="notTouch" role="menuitem" tabindex="0">
        Subtitles/CC
        <span class="spacer"></span>
        <span class="hint">
          None
        </span>
      </vime-menu-item>
      DefaultSlot
    </div>
  </vime-menu>
</vime-settings>
`;

exports[`should be structurally sound 1`] = `
<vime-player aria-hidden="true" aria-label="Video Player" id="vime-player-1" tabindex="0">
  <!---->
  <div class="active skeleton"></div>
  <vime-faketube></vime-faketube>
  <vime-ui>
    <!---->
    <vime-default-settings>
      <!---->
      <vime-settings style="bottom: 73px;">
        <!---->
        <vime-menu aria-hidden="true" aria-labelledby="" id="vime-settings-1" role="menu" tabindex="-1">
          <!---->
          <div>
            <vime-menu-item aria-hidden="false" aria-label="Playback Rate" class="notTouch" role="menuitem" tabindex="0">
              Playback Rate
              <span class="spacer"></span>
              <span class="hint">
                Normal
              </span>
            </vime-menu-item>
            <vime-menu-item aria-hidden="false" aria-label="Playback Quality" class="notTouch" role="menuitem" tabindex="0">
              Playback Quality
              <span class="spacer"></span>
              <span class="hint">
                Auto
              </span>
            </vime-menu-item>
            <vime-menu-item aria-hidden="false" aria-label="Subtitles/CC" class="notTouch" role="menuitem" tabindex="0">
              Subtitles/CC
              <span class="spacer"></span>
              <span class="hint">
                None
              </span>
            </vime-menu-item>
            DefaultSlot
          </div>
        </vime-menu>
      </vime-settings>
    </vime-default-settings>
  </vime-ui>
</vime-player>
`;
